src/clj/puppetlabs/http/client/async.clj

```diff
@@ -326,39 +137,81 @@
     (.start client)
     client))
 
-(schema/defn execute-with-consumer
-  [client :- HttpAsyncClient
-   request :- HttpRequest
-   future-callback :- FutureCallback
-   result :- common/ResponsePromise
-   opts :- common/RequestOptions
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Map the Ring request onto the Java API
+
+(schema/defn callback-response :- common/Response
+  [opts :- common/UserRequestOptions
    callback :- common/ResponseCallbackFn
-   http-context :- HttpContext]
-  (let [;; Create an Apache AsyncResponseConsumer that will return the response to us as soon as it is 
-        ;; available then send the response body asynchronously
-        consumer (StreamingAsyncResponseConsumer.
-                  (FnDeliverable.
-                   (fn
-                     [http-response]
-                     (complete-response result opts callback http-response http-context))))
-        ;; If an error occurs early in the request, the consumer may not get a chance to return the
-        ;; response using the FnDeliverable. This wrapper around the future-callback guarantees that
-        ;; happens. It also takes care of notifying the consumer of the final result.
-        streaming-complete-callback (reify
-                                      FutureCallback
-                                      (completed [this http-response]
-                                        (.setFinalResult consumer nil)
-                                        (.completed future-callback http-response))
-                                      (failed [this ex]
-                                        (.setFinalResult consumer ex)
-                                        (.failed future-callback ex))
-                                      (cancelled [this]
-                                        (.setFinalResult consumer nil)
-                                        (.cancelled future-callback)))]
-    (.execute client
-              (HttpAsyncMethods/create request)
-              consumer
-              streaming-complete-callback)))
+   response :- common/Response]
+  (if callback
+    (try
+      (callback response)
+      (catch Exception e
+        {:opts opts :error e}))
+    response))
+
+(schema/defn java-content-type->clj :- common/ContentType
+  [java-content-type]
+  (if java-content-type
```

Comment:
Added the test and updated the Java content type parsing to check for empty. In the last commit.